### PR TITLE
OperatorModel - cleanup and testing

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

This PR is for finalizing the design of the `OperatorModel` class and adding tests

### Update

This PR is ready for review. Contains:

- Main piece: `OperatorModel` class in `models.py`. 
     - As explained in the documentation, represents a list of operators with signal coefficients, with functionality for evaluating the model, entering a frame, and specifying a frequency cutoff (for things like the RWA). 
     - This file also contains a helper class called `FrameFreqHelper` whose only purpose is to compartmentalize the more technical aspects of the calculations to do with frames/frequencies in `OperatorModel`.
- Secondary piece: Addition of `VectorSignal` to `signals.py`. 
    - Internally, the `OperatorModel` represents the signals as a `VectorSignal` - a class that is the same as `Signal` but has a vector valued envelope and a list of carrier frequencies. 
    - The point of this is to fully abstract the calculation of signal values out of `OperatorModel`, to facilitate future applications in which it is computationally advantageous to compute the values of multiple signals at once (as opposed to iterating through in a list).
    - A `VectorSignal` can be instantiated from a list of `Signal` objects, and the `OperatorModel` will attempt to do this if the signals are passed as a list (preserving the current standard use-case of specifying signals as a list).
- Testing:
  - New test file `pulse_new/models/test_operator_models.py`, containing various tests for `OperatorModel`
- Demo
  - New demo notebook `OperatorModel_demo.ipynb` demonstrating constructing an `OperatorModel`, calling some functions, the effects of entering a frame, and the effects of specifying a frequency cutoff.

~~I think some minor further additions: add additional tests for handling of non-diagonal frame-bases (i.e. evaluation of a model in the basis which the frame is diagonal), as well as the related functionality in `FrameFreqHelper` for specifying which basis to express a "state" in when transforming into/out of the frame. Could add this stuff to the demo too.~~

In addition to the above:
- A bunch of documentation as per @eggerdj 's comments
- Added a check that a specified signal has the same number of entries as the operators
- Removed the function `vector_apply_diag_frame` and moved the code into the only function that was calling it
- Added a section to the `OperatorModel` demo for specifying a `signal_mapping`
- Added a section to `OperatorModel` demo showing helper functions for transforming states into/out of the given frame (which will be used in DE solvers)

Added tests:
- checking warning if length of signal is invalid
- test for setting a `signal_mapping`
- tests for state frame transformations

### CODE REVIEW

**Comment by DanPuzzuoli** [comment]
Current status: renamed file `models.py` to `operator_models.py` and the `BaseModel` class to `OperatorModel` as the original names seemed too general.

**Comment by DanPuzzuoli** [comment]
Adding a bunch of tests for the `FrameFreqHelper`. Tests that remain:
- `evaluate` with frequency cutoffs
- frame transformations for states
- both of the above with various setting of flags for whether the input/output of the function is in the basis in which the frame is diagonal

**Comment by DanPuzzuoli** [comment]
This is ready for review (see updated initial comment for details).

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                 signals,
+                 signal_mapping=None,
+                 frame_operator=None,
+                 cutoff_freq=None):
+        """Initialize.
+
+        Args:
+            operators (list): list of Operator objects.
+            signals (list): Specifiable as either a VectorSignal, a list of
+                            Signal objects, or as the inputs to signal_mapping.
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: Perhaps we can add some type hints in the signature of the functions to represent this.
E.g. `Union[List, VectorSignal]`?

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+            signal_mapping (Callable): a function returning wither a
+                                       VectorSignal or a list of Signal objects.
+
+            frame_operator (Operator): Rotating frame operator.
+            cutoff_freq (float): Frequency cutoff when evaluating the model.
+        """
+
+        self._operators = operators
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py:93]: Should we check that operators and signals have the same length?

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+                 frame_operator=None,
+                 cutoff_freq=None):
+        """Initialize.
+
+        Args:
+            operators (list): list of Operator objects.
+            signals (list): Specifiable as either a VectorSignal, a list of
+                            Signal objects, or as the inputs to signal_mapping.
+            signal_mapping (Callable): a function returning wither a
+                                       VectorSignal or a list of Signal objects.
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: This docstring could greatly benefit from an example. Looking at the signals setter below it is not quite clear how this mapping should work.

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            self._signals = signals
+
+    @property
+    def frame_operator(self):
+        """Return the frame operator."""
+        return self._frame_operator
+
+    @frame_operator.setter
+    def frame_operator(self, frame_operator):
+        """Set the frame operator; must be anti-Hermitian.
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: Can we spend a few more lines on what the frame operator is and what it does?

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+            - An Operator object
+            - A 2d array
+            - A 1d array - in which case it is assumed the frame operator is a
+              diagonal matrix, and the array gives the diagonal elements.
+        """
+
+        if isinstance(frame_operator, np.ndarray) and frame_operator.ndim == 1:
+            # if 1d array check purely imaginary
+            if np.linalg.norm(frame_operator + frame_operator.conj()) > 10**-10:
+                raise Exception('frame_op must correspond to an anti-Hermitian matrix.')
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: Does `Operator` not work on 1D arrays?

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+        Args:
+            time: Time at which to create the generator.
+            y: operator or vector to apply the model to.
+            in_frame_diag_basis: whether to evaluate the frame in the frame basis
+        """
+        return np.dot(y, self.evaluate(time, in_frame_basis))
+
+    @property
+    def drift(self):
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]:
```suggestion
    def drift(self) -> np.array:
```

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                                       object or a 1d array (in which case it
+                                       is assumed to already be diagonalized.)
+
+                                       the frame_operator is assumed to be
+                                       anti-hermitian.
+            cutoff_freq (float): Cutoff frequency when evaluating generator.
+                                 If None, no cutoff is performed.
+        """
+
+        # initial setup of frame operator
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: redundent?
```suggestion
```

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        D_diff = D_diff - D_diff.transpose()
+
+        # set up matrix encoding frequencies
+        im_angular_freqs = 1j * 2 * np.pi * self.carrier_freqs
+        self._S = np.array([w + D_diff for w in im_angular_freqs])
+
+        self._M_cutoff = None
+        if cutoff_freq is not None:
+            self._M_cutoff = ((np.abs(self._S.imag) / (2 * np.pi)) <
+                                            self.cutoff_freq).astype(int)
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: Can we add a comment what this is for?

File: qiskit/providers/aer/pulse_new/models/operator_models.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+
+        if return_in_frame_basis:
+            return out_in_fb
+        else:
+            return self.frame_basis @ out_in_fb
+
+
+def vector_apply_diag_frame(t, mats_in_frame_basis, coeffs, S, M_cutoff=None):
+    """Given a list of matrices specified in the frame_basis for a
+    """
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/operator_models.py]: Incomplete docstring

File: qiskit/providers/aer/pulse_new/models/signals.py [hunk 10]
```diff
@@ -211,3 +212,96 @@ def signal_add(sig1, sig2):
...
+    def __init__(self,
+                 envelope,
+                 carrier_freqs,
+                 drift_array=None):
+        """Initialize.
+
+        Args:
+            envelope (Callable): function of a single float returning an array.
+            carrier_freqs (array): list of carrier frequences
+            drift_array (array): the drift array
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/signals.py]: Needs more information.

**Review by eggerdj — CHANGES_REQUESTED** [review]
I think this is looking really good. A few comments:
* In some places a bit more detail in the docstrings would help.
* Some places are missing type hints.

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 5 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py]: Currently it doesn't

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 1 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py]: Sounds good - going to add these type hints everywhere to be consistent.

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 4 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py]: I added a sentence pointing to the class doc string for this, where the effects of setting a frame operator are generally explained.

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 9 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py]: I've eliminated this function and just moved the computations into the `FrameFreqHelper.evaluate` method - it was too weird to explain out of context.

File: qiskit/providers/aer/pulse_new/models/signals.py (hunk 10 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/signals.py]: Tried adding some here - there is also more info in the class doc string

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 3 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py]: I've added an example into the demo as well as in the class doc string for how this is meant to behave.

File: qiskit/providers/aer/pulse_new/models/operator_models.py (hunk 2 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/pulse_new/models/operator_models.py:93]: Hmm not sure... I think how to handle length checking depends on what type of functionality we want to have for adding terms to a model. That being said we don't have any of that functionality, so perhaps for now I can just add a check when the signals are set that they're the same length as the operators (with the operators being fixed at instantiation).

**Comment by DanPuzzuoli** [comment]
Summary of changes:
- A bunch of documentation as per @eggerdj 's comments
- Added a check that a specified signal has the same number of entries as the operators
- Removed the function `vector_apply_diag_frame` and moved the code into the only function that was calling it
- Added a section to the `OperatorModel` demo for specifying a `signal_mapping`

Added tests:
- checking warning if length of signal is invalid
- test for setting a `signal_mapping`

File: qiskit/providers/aer/pulse_new/models/signals.py [hunk 11]
```diff
@@ -374,3 +374,106 @@ def signal_add(sig1: Union[BaseSignal, float, int, complex],
...
+
+        # if not supplied nothing is assumed, constant array is taken as all
+        # zeros
+        if drift_array is None:
+            self.drift_array = np.zeros(len(self.carrier_freqs))
+        else:
+            self.drift_array = drift_array
+
+    @classmethod
+    def from_signal_list(cls, signal_list: List[Signal]):
```
Comment (eggerdj) [inline qiskit/providers/aer/pulse_new/models/signals.py]: Can we initialize from `BaseSignal` instead? `BaseSignal` defines an `envelop_value` function so this should be fine.
```suggestion
    def from_signal_list(cls, signal_list: List[BaseSignal]):
```

**Review by eggerdj — APPROVED** [review]
Looks good, one minor comment on `VectorSignal`.

**Pull request merged**